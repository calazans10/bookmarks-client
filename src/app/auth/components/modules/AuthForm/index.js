import React from 'react';
import PropTypes from 'prop-types';
import { Form, Field } from 'react-final-form';
import MainForm from '../../../../core/components/modules/MainForm';
import FormGroup from '../../../../core/components/modules/FormGroup';
import ButtonSubmit from '../../../../core/components/modules/ButtonSubmit';
import {
  required,
  mustBeFullName,
  mustBeEmail,
  composeValidators,
} from '../../../../../utils/validators';

const AuthForm = ({ legend, action, isRegistration, onSubmit }) => (
  <Form
    onSubmit={onSubmit}
    render={({ handleSubmit }) => (
      <MainForm legend={legend} onSubmit={handleSubmit}>
        {isRegistration && (
          <Field
            name="name"
            validate={composeValidators(required, mustBeFullName)}
            render={({ input, meta }) => (
              <FormGroup input={input} meta={meta} label="Name" type="text" />
            )}
          />
        )}
        <Field
          name="email"
          validate={composeValidators(required, mustBeEmail)}
          render={({ input, meta }) => (
            <FormGroup input={input} meta={meta} label="Email" type="email" />
          )}
        />
        <Field
          name="password"
          validate={required}
          render={({ input, meta }) => (
            <FormGroup input={input} meta={meta} label="Password" type="password" />
          )}
        />
        <div>
          <ButtonSubmit>{action}</ButtonSubmit>
        </div>
      </MainForm>
    )}
  />
);

AuthForm.propTypes = {
  legend: PropTypes.string.isRequired,
  action: PropTypes.string.isRequired,
  isRegistration: PropTypes.bool,
  onSubmit: PropTypes.func.isRequired,
};

AuthForm.defaultProps = {
  isRegistration: false,
};

export default AuthForm;
