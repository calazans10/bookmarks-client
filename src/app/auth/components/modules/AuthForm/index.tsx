import React from 'react';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers';
import * as yup from 'yup';
import { ErrorMessages } from 'enums';
import { AuthData } from 'app/auth/types';
import MainForm from 'app/core/components/modules/MainForm';
import FormGroup from 'app/core/components/modules/FormGroup';
import ButtonSubmit from 'app/core/components/modules/ButtonSubmit';

type AuthFormProps = {
  legend: string;
  action: string;
  isRegistration: boolean;
  onSubmit: (values: AuthData) => Promise<void>;
} & typeof defaultProps;

const defaultProps = {
  isRegistration: false,
};

const schema = yup.object().shape({
  firstName: yup.string().when('isRegistration', {
    is: 'YES',
    then: yup.string().required(ErrorMessages.REQUIRED),
  }),
  lastName: yup.string().when('isRegistration', {
    is: 'YES',
    then: yup.string().required(ErrorMessages.REQUIRED),
  }),
  email: yup.string().email(ErrorMessages.EMAIL).required(ErrorMessages.REQUIRED),
  password: yup.string().required(ErrorMessages.REQUIRED),
});

const AuthForm = ({ legend, action, isRegistration, onSubmit }: AuthFormProps) => {
  const { register, handleSubmit, errors } = useForm<AuthData>({
    resolver: yupResolver(schema),
  });

  return (
    <MainForm legend={legend} onSubmit={handleSubmit(onSubmit)}>
      <input
        name="isRegistration"
        type="hidden"
        defaultValue={isRegistration ? 'YES' : 'NO'}
        ref={register}
      />
      {isRegistration && (
        <>
          <FormGroup
            name="firstName"
            label="First Name"
            type="text"
            inputRef={register}
            error={errors.firstName?.message}
          />
          <FormGroup
            name="lastName"
            label="Last Name"
            type="text"
            inputRef={register}
            error={errors.lastName?.message}
          />
        </>
      )}
      <FormGroup
        name="email"
        label="E-mail"
        type="email"
        inputRef={register}
        error={errors.email?.message}
      />
      <FormGroup
        name="password"
        label="Password"
        type="password"
        inputRef={register}
        error={errors.password?.message}
      />
      <div>
        <ButtonSubmit>{action}</ButtonSubmit>
      </div>
    </MainForm>
  );
};

AuthForm.defaultProps = defaultProps;

export default AuthForm;
