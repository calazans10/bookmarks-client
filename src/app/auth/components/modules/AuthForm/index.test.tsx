import React from 'react';
import { render } from '@testing-library/react';
import AuthForm from './index';

describe('AuthForm', () => {
  describe('when is registration', () => {
    it('renders without crashing', () => {
      const { container } = render(
        <AuthForm legend="A legend" action="Save" isRegistration onSubmit={jest.fn()} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('when is not registration', () => {
    it('renders without crashing', () => {
      const { container } = render(
        <AuthForm legend="A legend" action="Save" isRegistration={false} onSubmit={jest.fn()} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
