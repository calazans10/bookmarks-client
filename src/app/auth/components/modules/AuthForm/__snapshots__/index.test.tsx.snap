// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm when is not registration renders without crashing 1`] = `
<form
  autocomplete="off"
  class="sc-AxjAm hYYXIy"
  novalidate=""
>
  <fieldset>
    <legend
      class="sc-AxirZ dBtpcc"
    >
      A legend
    </legend>
    <input
      name="isRegistration"
      type="hidden"
      value="NO"
    />
    <div
      class="sc-AxiKw bLDZyO"
    >
      <label
        class="sc-AxhCb ieTFEo"
        for="email"
      >
        E-mail
      </label>
      <input
        aria-invalid="false"
        class="sc-AxhUy ePtFHq"
        id="email"
        name="email"
        placeholder="E-mail"
        type="email"
        value=""
      />
      
    </div>
    <div
      class="sc-AxiKw bLDZyO"
    >
      <label
        class="sc-AxhCb ieTFEo"
        for="password"
      >
        Password
      </label>
      <input
        aria-invalid="false"
        class="sc-AxhUy ePtFHq"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      
    </div>
    <div>
      <button
        class="sc-AxheI dSQsuG"
        type="submit"
      >
        Save
      </button>
    </div>
  </fieldset>
</form>
`;

exports[`AuthForm when is registration renders without crashing 1`] = `
<form
  autocomplete="off"
  class="sc-AxjAm hYYXIy"
  novalidate=""
>
  <fieldset>
    <legend
      class="sc-AxirZ dBtpcc"
    >
      A legend
    </legend>
    <input
      name="isRegistration"
      type="hidden"
      value="YES"
    />
    <div
      class="sc-AxiKw bLDZyO"
    >
      <label
        class="sc-AxhCb ieTFEo"
        for="firstName"
      >
        First Name
      </label>
      <input
        aria-invalid="false"
        class="sc-AxhUy ePtFHq"
        id="firstName"
        name="firstName"
        placeholder="First Name"
        type="text"
        value=""
      />
      
    </div>
    <div
      class="sc-AxiKw bLDZyO"
    >
      <label
        class="sc-AxhCb ieTFEo"
        for="lastName"
      >
        Last Name
      </label>
      <input
        aria-invalid="false"
        class="sc-AxhUy ePtFHq"
        id="lastName"
        name="lastName"
        placeholder="Last Name"
        type="text"
        value=""
      />
      
    </div>
    <div
      class="sc-AxiKw bLDZyO"
    >
      <label
        class="sc-AxhCb ieTFEo"
        for="email"
      >
        E-mail
      </label>
      <input
        aria-invalid="false"
        class="sc-AxhUy ePtFHq"
        id="email"
        name="email"
        placeholder="E-mail"
        type="email"
        value=""
      />
      
    </div>
    <div
      class="sc-AxiKw bLDZyO"
    >
      <label
        class="sc-AxhCb ieTFEo"
        for="password"
      >
        Password
      </label>
      <input
        aria-invalid="false"
        class="sc-AxhUy ePtFHq"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      
    </div>
    <div>
      <button
        class="sc-AxheI dSQsuG"
        type="submit"
      >
        Save
      </button>
    </div>
  </fieldset>
</form>
`;
